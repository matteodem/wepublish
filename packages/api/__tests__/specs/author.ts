import {MongoDBAdapter} from '@wepublish/api-db-mongodb'
import {ApolloServerTestClient} from 'apollo-server-testing'
import {createGraphQLTestClientWithMongoDB} from '../utility'
import {CreateAuthor, AuthorInput, AuthorList} from '../api/private'

let testClientPublic: ApolloServerTestClient
let testClientPrivate: ApolloServerTestClient
let dbAdapter: MongoDBAdapter

beforeAll(async () => {
  try {
    const setupClient = await createGraphQLTestClientWithMongoDB()
    testClientPublic = setupClient.testClientPublic
    testClientPrivate = setupClient.testClientPrivate
    dbAdapter = setupClient.dbAdapter

    console.log('public', testClientPublic)
  } catch (error) {
    console.log('Error', error)
    throw new Error('Error during test setup')
  }
})

describe('Authors', () => {
  describe('can be created/edited/deleted:', () => {
    let authorId = ''
    beforeEach(async () => {
      const {mutate} = testClientPrivate
      const authorInput: AuthorInput = {
        name: 'JRR Tolkien',
        slug: 'j-tolkien',
        bio: []
      }
      const res = await mutate({
        mutation: CreateAuthor,
        variables: {
          input: authorInput
        }
      })
      authorId = res.data?.createAuthor?.id
      console.log(authorId)
    })

    test('can be created', async () => {
      const {mutate} = testClientPrivate
      const authorInput: AuthorInput = {
        name: 'John Grisham',
        slug: 'j-grish',
        bio: []
      }
      const res = await mutate({
        mutation: CreateAuthor,
        variables: {
          input: authorInput
        }
      })
      expect(res).toMatchSnapshot({
        data: {
          createAuthor: {
            id: expect.any(String)
          }
        }
      })
    })

    test('can be read in list', async () => {
      const {query} = testClientPrivate
      const res = await query({
        query: AuthorList,
        variables: {
          first: 5
        }
      })
      expect(res).toMatchSnapshot({
        data: {
          authors: {
            nodes: expect.any(Array),
            pageInfo: {
              endCursor: expect.any(String),
              startCursor: expect.any(String)
            }
            //totalCount will return size of list which changes depending on how many tests were run
            //expect any number?
          }
        }
      })
    })
  })
})

afterAll(async () => {
  if (dbAdapter) {
    await dbAdapter.db.dropDatabase()
    await dbAdapter.client.close()
  }
})
