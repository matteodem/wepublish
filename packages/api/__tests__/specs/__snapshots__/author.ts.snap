// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authors can be created/edited/deleted: can be created 1`] = `
Object {
  "data": Object {
    "createAuthor": Object {
      "bio": Array [],
      "id": Any<String>,
      "image": null,
      "links": null,
      "name": "John Grisham",
      "slug": "j-grish",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Authors can be created/edited/deleted: can be read in list 1`] = `
Object {
  "data": Object {
    "authors": Object {
      "nodes": Any<Array>,
      "pageInfo": Object {
        "endCursor": Any<String>,
        "hasNextPage": false,
        "hasPreviousPage": false,
        "startCursor": Any<String>,
      },
      "totalCount": 2,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
